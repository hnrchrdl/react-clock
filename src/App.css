section {
  display: flex;
  justify-content: center;
  margin: 10px;
  padding: 10px;
  border-bottom: 1px solid #eee;
}
section > * {
  min-width: 250px;
  margin: 10px;
  padding: 10px;
}

.button {
  font-size: 0.7em;
  padding: 10px;
  cursor: pointer;
  color: #999;
}
.button:hover {
  color: #111;
}
.button.active {
  color: coral;
}
.active {
  color: coral;
}
