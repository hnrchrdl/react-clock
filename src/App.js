// @flow

import React, { Component } from 'react';
import './App.css';
import { minToSec, friendlyTime } from './lib';
import Button from './Button';

type Props = {};

type State = {
  break: number,
  session: number,
  time: number,
  touched: boolean,
  paused: boolean,
  isSession: boolean
};

const initialState: State = {
  break: minToSec(5),
  session: minToSec(25),
  time: minToSec(25),
  touched: false,
  paused: false,
  isSession: true
};

class App extends Component<Props, State> {
  state = initialState;
  interval: any;
  decrementTimer = () => {
    this.setState(
      state => ({ time: state.time - 1 }),
      () => {
        if (this.state.time === 0) {
          this.setState(state => ({
            time: state.isSession ? state.break : state.session,
            isSession: !state.isSession
          }));
        }
      }
    );
  };
  incrementBreak = () => {
    this.setState(state => ({ break: state.break + minToSec(1) }));
  };
  decrementBreak = () => {
    this.setState(state => ({ break: state.break - minToSec(1) }));
  };
  incrementSession = () => {
    this.setState(state => ({ session: state.session + minToSec(1) }));
  };
  decrementSession = () => {
    this.setState(state => ({ session: state.session - minToSec(1) }));
  };
  start = () => {
    if (!this.interval) {
      this.setState({ touched: true, paused: false });
      this.interval = setInterval(this.decrementTimer, 1000);
    }
  };
  pause = () => {
    this.setState({ paused: true });
    clearInterval(this.interval);
    this.interval = null;
  };
  stop = () => {
    clearInterval(this.interval);
    this.interval = null;
    this.setState(state => ({
      ...initialState,
      break: state.break,
      session: state.session,
      time: state.session
    }));
  };
  render() {
    return (
      <div className="App">
        <section>
          <div className="title">Pomodoro Clock</div>
        </section>
        <section>
          <div className={`${this.state.isSession ? '' : 'active'}`}>
            <div>BREAK</div>
            {friendlyTime(this.state.break)}
            <div>
              <Button onClick={this.incrementBreak} text="+" />
              <Button onClick={this.decrementBreak} text="-" />
            </div>
          </div>
          <div className={`${this.state.isSession ? 'active' : ''}`}>
            <div>SESSION</div>
            {friendlyTime(this.state.session)}
            <div>
              <Button onClick={this.incrementSession} text="+" />
              <Button onClick={this.decrementSession} text="-" />
            </div>
          </div>
        </section>
        <section>
          <div>
            {friendlyTime(this.state.time)}
            <div>
              <Button
                onClick={this.start}
                text="START"
                active={this.state.touched && !this.state.paused}
              />
              <Button
                onClick={this.pause}
                text="PAUSE"
                active={this.state.touched && this.state.paused}
              />
              <Button
                onClick={this.stop}
                text="RESET"
                active={!this.state.touched}
              />
            </div>
          </div>
        </section>
      </div>
    );
  }
}

export default App;
