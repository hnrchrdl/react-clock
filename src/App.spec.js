// @flow

import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

it('should match snapshot', () => {
  const wrapper: App = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});
