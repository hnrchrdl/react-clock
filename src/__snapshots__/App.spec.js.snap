// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <section>
          <div
                    className="title"
          >
                    Pomodoro Clock
          </div>
</section>,
        <section>
          <div
                    className=""
          >
                    <div>
                              BREAK
                    </div>
                    05:00
                    <div>
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="+"
                              />
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="-"
                              />
                    </div>
          </div>
          <div
                    className="active"
          >
                    <div>
                              SESSION
                    </div>
                    25:00
                    <div>
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="+"
                              />
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="-"
                              />
                    </div>
          </div>
</section>,
        <section>
          <div>
                    25:00
                    <div>
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="START"
                              />
                              <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="PAUSE"
                              />
                              <Button
                                        active={true}
                                        onClick={[Function]}
                                        text="RESET"
                              />
                    </div>
          </div>
</section>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="title"
>
            Pomodoro Clock
</div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pomodoro Clock",
            "className": "title",
          },
          "ref": null,
          "rendered": "Pomodoro Clock",
          "type": "div",
        },
        "type": "section",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className=""
>
              <div>
                            BREAK
              </div>
              05:00
              <div>
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="+"
                            />
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="-"
                            />
              </div>
</div>,
            <div
              className="active"
>
              <div>
                            SESSION
              </div>
              25:00
              <div>
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="+"
                            />
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="-"
                            />
              </div>
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div>
                  BREAK
</div>,
                "05:00",
                <div>
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="+"
                  />
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="-"
                  />
</div>,
              ],
              "className": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "BREAK",
                },
                "ref": null,
                "rendered": "BREAK",
                "type": "div",
              },
              "05:00",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="+"
/>,
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="-"
/>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "+",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "-",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div>
                  SESSION
</div>,
                "25:00",
                <div>
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="+"
                  />
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="-"
                  />
</div>,
              ],
              "className": "active",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "SESSION",
                },
                "ref": null,
                "rendered": "SESSION",
                "type": "div",
              },
              "25:00",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="+"
/>,
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="-"
/>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "+",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "-",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "section",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            25:00
            <div>
                        <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="START"
                        />
                        <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="PAUSE"
                        />
                        <Button
                                    active={true}
                                    onClick={[Function]}
                                    text="RESET"
                        />
            </div>
</div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "25:00",
              <div>
                <Button
                                active={false}
                                onClick={[Function]}
                                text="START"
                />
                <Button
                                active={false}
                                onClick={[Function]}
                                text="PAUSE"
                />
                <Button
                                active={true}
                                onClick={[Function]}
                                text="RESET"
                />
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "25:00",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Button
                    active={false}
                    onClick={[Function]}
                    text="START"
/>,
                  <Button
                    active={false}
                    onClick={[Function]}
                    text="PAUSE"
/>,
                  <Button
                    active={true}
                    onClick={[Function]}
                    text="RESET"
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "active": false,
                    "onClick": [Function],
                    "text": "START",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "active": false,
                    "onClick": [Function],
                    "text": "PAUSE",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "active": true,
                    "onClick": [Function],
                    "text": "RESET",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "section",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <section>
            <div
                        className="title"
            >
                        Pomodoro Clock
            </div>
</section>,
          <section>
            <div
                        className=""
            >
                        <div>
                                    BREAK
                        </div>
                        05:00
                        <div>
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="+"
                                    />
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="-"
                                    />
                        </div>
            </div>
            <div
                        className="active"
            >
                        <div>
                                    SESSION
                        </div>
                        25:00
                        <div>
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="+"
                                    />
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="-"
                                    />
                        </div>
            </div>
</section>,
          <section>
            <div>
                        25:00
                        <div>
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="START"
                                    />
                                    <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="PAUSE"
                                    />
                                    <Button
                                                active={true}
                                                onClick={[Function]}
                                                text="RESET"
                                    />
                        </div>
            </div>
</section>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="title"
>
              Pomodoro Clock
</div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Pomodoro Clock",
              "className": "title",
            },
            "ref": null,
            "rendered": "Pomodoro Clock",
            "type": "div",
          },
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className=""
>
                <div>
                                BREAK
                </div>
                05:00
                <div>
                                <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="+"
                                />
                                <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="-"
                                />
                </div>
</div>,
              <div
                className="active"
>
                <div>
                                SESSION
                </div>
                25:00
                <div>
                                <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="+"
                                />
                                <Button
                                                active={false}
                                                onClick={[Function]}
                                                text="-"
                                />
                </div>
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div>
                    BREAK
</div>,
                  "05:00",
                  <div>
                    <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="+"
                    />
                    <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="-"
                    />
</div>,
                ],
                "className": "",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "BREAK",
                  },
                  "ref": null,
                  "rendered": "BREAK",
                  "type": "div",
                },
                "05:00",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <Button
                        active={false}
                        onClick={[Function]}
                        text="+"
/>,
                      <Button
                        active={false}
                        onClick={[Function]}
                        text="-"
/>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "active": false,
                        "onClick": [Function],
                        "text": "+",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "active": false,
                        "onClick": [Function],
                        "text": "-",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div>
                    SESSION
</div>,
                  "25:00",
                  <div>
                    <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="+"
                    />
                    <Button
                                        active={false}
                                        onClick={[Function]}
                                        text="-"
                    />
</div>,
                ],
                "className": "active",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "SESSION",
                  },
                  "ref": null,
                  "rendered": "SESSION",
                  "type": "div",
                },
                "25:00",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <Button
                        active={false}
                        onClick={[Function]}
                        text="+"
/>,
                      <Button
                        active={false}
                        onClick={[Function]}
                        text="-"
/>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "active": false,
                        "onClick": [Function],
                        "text": "+",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "active": false,
                        "onClick": [Function],
                        "text": "-",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
          ],
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              25:00
              <div>
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="START"
                            />
                            <Button
                                          active={false}
                                          onClick={[Function]}
                                          text="PAUSE"
                            />
                            <Button
                                          active={true}
                                          onClick={[Function]}
                                          text="RESET"
                            />
              </div>
</div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "25:00",
                <div>
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="START"
                  />
                  <Button
                                    active={false}
                                    onClick={[Function]}
                                    text="PAUSE"
                  />
                  <Button
                                    active={true}
                                    onClick={[Function]}
                                    text="RESET"
                  />
</div>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "25:00",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="START"
/>,
                    <Button
                      active={false}
                      onClick={[Function]}
                      text="PAUSE"
/>,
                    <Button
                      active={true}
                      onClick={[Function]}
                      text="RESET"
/>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "START",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": false,
                      "onClick": [Function],
                      "text": "PAUSE",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "active": true,
                      "onClick": [Function],
                      "text": "RESET",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "section",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
